tempesta_fw/http_limits.c

```diff
@@ -638,70 +635,31 @@ frang_http_host_check(const TfwHttpReq *req, FrangAcc *ra)
 	return ret;
 }
 
-/**
- * Monotonically increasing time quantums. The configured @tframe
- * is divided by FRANG_FREQ slots to get the quantums granularity.
- */
-static unsigned int
-frang_resp_quantum(unsigned short tframe)
-{
-	return jiffies * FRANG_FREQ / (tframe * HZ);
-}
-
-static int
-frang_bad_resp_limit(FrangAcc *ra, FrangHttpRespCodeBlock *resp_cblk)
-{
-	FrangRespCodeStat *stat = ra->resp_code_stat;
-	unsigned long cnt = 0;
-	const unsigned int ts = frang_resp_quantum(resp_cblk->tf);
-	int i = 0;
-
-	for (; i < FRANG_FREQ; ++i) {
-		if (frang_time_in_frame(ts, stat[i].ts))
-			cnt += stat[i].cnt;
-	}
-	if (cnt > resp_cblk->limit) {
-		frang_limmsg("http_resp_code_block limit", cnt,
-			     resp_cblk->limit, &FRANG_ACC2CLI(ra)->addr);
-		return TFW_BLOCK;
-	}
-	return TFW_PASS;
-}
-
 /*
  * The GFSM states aren't hookable, so don't open the states definitions and
  * only start and finish states are present.
  */
 #define TFW_GFSM_FRANG_REQ_STATE(s)					\
 	((TFW_FSM_FRANG_REQ << TFW_GFSM_FSM_SHIFT) | (s))
 enum {
+	/* Initial Frang Req state, not hookable */
 	TFW_FRANG_REQ_FSM_INIT	= TFW_GFSM_FRANG_REQ_STATE(0),
-	TFW_FRANG_REQ_FSM_DONE	= TFW_GFSM_FRANG_REQ_STATE(TFW_GFSM_STATE_LAST)
-};
-
-enum {
-	Frang_Req_0 = 0,
-
-	Frang_Req_Hdr_Start,
-	Frang_Req_Hdr_Method,
-	Frang_Req_Hdr_UriLen,
-	Frang_Req_Hdr_FieldDup,
-	Frang_Req_Hdr_FieldLen,
-	Frang_Req_Hdr_FieldLenFinal,
-	Frang_Req_Hdr_Crlf,
-	Frang_Req_Hdr_Host,
-	Frang_Req_Hdr_ContentType,
+	/* New request is allowed on the client connection. */
+	TFW_FRANG_REQ_NEW_ALLOWED = TFW_GFSM_FRANG_REQ_STATE(1),
+	/* Request method is allowed. */
+	TFW_FRANG_REQ_METH_ALLOWED = TFW_GFSM_FRANG_REQ_STATE(2),
+	/* Request uri is allowed. */
+	TFW_FRANG_REQ_URI_ALLOWED = TFW_GFSM_FRANG_REQ_STATE(3),
+	/* Request headers are not parsed fully. */
+	TFW_FRANG_REQ_HDRS_INCOMPLETE = TFW_GFSM_FRANG_REQ_STATE(4),
+	/* Request headers are validated. */
+	TFW_FRANG_REQ_HDRS_ALLOWED = TFW_GFSM_FRANG_REQ_STATE(5),
+	/* Request body is not parsed fully. */
+	TFW_FRANG_REQ_BODY_INCOMPLETE = TFW_GFSM_FRANG_REQ_STATE(6),
+	/* Full body is valid, but some final checks are to be performed. */
+	TFW_FRANG_REQ_BODY_ALLOWED = TFW_GFSM_FRANG_REQ_STATE(7),
 
-	Frang_Req_Hdr_NoState,
-
-	Frang_Req_Body_Start,
-	Frang_Req_Body_Timeout,
-	Frang_Req_Body_ChunkCnt,
-	Frang_Req_Body_Len,
-
-	Frang_Req_Body_NoState,
-
-	Frang_Req_Done
+	TFW_FRANG_REQ_FSM_DONE	= TFW_GFSM_FRANG_REQ_STATE(TFW_GFSM_STATE_LAST)
 };
```

Comment:
Yes, there are frang fsm cycles, I  didn't noticed that when created the issue. But at this time I changed the current code, because I don't like it.

First of all it defines FSM inside FSM. The top level frang FSM states (`TFW_FRANG_REQ_FSM_INIT`|`TFW_FRANG_REQ_FSM_DONE`) are stored in `conn->state`. But frang also keeps per-request states in `req->frang_st`. I don't see any reasons, why need to distinct those top-level states from the per-request states, and define two FSMs instead of one.

Then, `Frang_Req_Done` state is reached only when all body checks are disabled. It's never reached, if at least one body check is enabled.

One check is done twice. Imagine, we have stopped on validating request in the middle of the headers, and now a new request chunk is parsed and frang callback is called. All headers was parsed fully this time. We move between FSM states in the following route: `Frang_Req_Hdr_FieldDup` -> `Frang_Req_Hdr_FieldLen` -> `Frang_Req_Hdr_Crlf` -> `Frang_Req_Hdr_FieldLenFinal` -> `Frang_Req_Hdr_Host` -> `Frang_Req_Hdr_ContentType` -> `Frang_Req_Body_Start`. The states `Frang_Req_Hdr_FieldLen` and `Frang_Req_Hdr_FieldLenFinal` are absolutely the same.

Trailer headers validation is missed.

So I've refactored frang FSM, not it doesn't contain cycles, and there is only one FSM, not two, and missing checks was added. As for me, frang became more straight forward now,